### PR TITLE
Adonis upgrade

### LINKED ISSUE #45: Upgrade Adonis
It seems like some components of adonis are heavily out-of-date preventing a normal production build. (See getferdi/server-docker#8)

### PR DESCRIPTION
**Pull request by cromefire**
Fixes getferdi/sponsors-only#1640 

The upgrade seemingly takes a bit more effort, will probably simplify some things and is also most likely a breaking change.

### CODE REVIEW

**Comment by cromefire** [comment]
Also switched the package manager to yarn, because I don't like waiting...

**Comment by cromefire** [comment]
The migration might have to [wait for auth to be done](https://preview.adonisjs.com/blog/introducing-adonisjs-v5/) though, but most stuff can probably already be done.

**Comment by mhatvan** [comment]
Hey @cromefire, this is awesome work! Any chance that you can finish up the PR?

In the meantime, some dependencies have newer major versions as well unfortunately, e.g. `@adonisjs/shield`. 
So you might want to upgrade again where necessary.

**Comment by cromefire** [comment]
I'll have a look at it in a few days, if I don't forget it.

**Comment by cromefire** [comment]
So I think we have a major problem, that we don't really get around: JWT auth isn't available anymore. That will most likely require restructuring in the client too.

**Comment by vantezzen** [comment]
> JWT auth isn't available anymore

Ferdi still uses JWT to stay compatible with Franz Servers. I'm not too familiar with the new Adonis update yet but as far as I can tell, we should be able to develop a custom JWT auth guard (or maybe there is already one available?) for JWT tokens, right?

**Comment by cromefire** [comment]
> Ferdi still uses JWT to stay compatible with Franz Servers. I'm not too familiar with the new Adonis update yet but as far as I can tell, we should be able to develop a custom JWT `guards.driver ` (or maybe there is already one available?) for JWT tokens, right?

Is the actual JWT actually used or is it just handled as an opaque token? If the latter we could just use the opaque token mechanism

**Comment by vantezzen** [comment]
Ferdi actually tries to decode the token to find out about the expiry date etc:

https://github.com/getferdi/ferdi/blob/a94ed659846f0bc93c05bccf0c96785b91e5c78f/src/stores/UserStore.js#L400-L406

**Comment by cromefire** [comment]
Well I'm not particularly keen on developing a JWT middleware myself, so from my point of view we either add a compatibility layer or find someone willing to develop an extension for it.

**Comment by cromefire** [comment]
Also, shouldn't this line be configurable:

> `Sentry.init({ dsn: 'https://34e9a42c1de24048b7bfc980211dd7c8@sentry.io/1838449' });`

**Comment by vantezzen** [comment]
Using a compatibility layer and normal tokens instead might be the better solution as it would make features like #23 possible that were previously not possible.
Also, thank you for taking the time to develop this upgrade so far!

**Comment by cromefire** [comment]
Another problem is adonisjs/persona#43 as persona is currently used and I don't see any way of upgrading without an upgrade of persona.

**Comment by mhatvan** [comment]
This is a very important PR to have an up-to-date infrastructure, is there a way to resolve the outstanding issues? I don't have knowledge with AdonisJS, but if I can help, let me know.

**Comment by cromefire** [comment]
Well I don't have much of an idea either, I mostly just wing it and try to fix what's not working, translate everything that's left to typescript. Mostly just takes quite a lot of time. If you want to help I can give you access to the branch.

**Comment by mhatvan** [comment]
Yeah I understand!

Can you summarize in bullet points maybe what still needs to be done?

This is also a blocker for https://github.com/getferdi/ferdi/pull/1783 unfortunately.

**Comment by cromefire** [comment]
Basically mostly translating anything that's left to typescript and then trying to get it to not error on startup.

**Comment by cromefire** [comment]
You should be able to write to the repo now. I'll also continue to work on this project, but it'll take a few weeks for me to have some more time again.

**Comment by mhatvan** [comment]
What I did:

- upgraded all deps to latest
- updated eslint, prettier, tsconfig
- switched from yarn to pnpm to be consistent with whole application
- removed unused test script
- renamed format to lint:fix script
- satisfied missing peer dependency warnings
- added `@types` for atob, btoa, node-fetch, uuid

Sorry in case I broke something @cromefire.

In the next few days, I will take time to convert remaining JS to TS and figure out Adonis and get it running hopefully ;)

I will also need APP_KEY and other required en vars to be able to launch the app...

**Comment by cromefire** [comment]
I think we can stick to yarn, it has great support across the ecosystem and basically already does everything that pnpm does for or use-case. Also could you please stick to the default eslint and prettier configs? They are already decent defaults and at least regarding the formatting part I'd like to keep it that way.

**Comment by cromefire** [comment]
The app key is something you can just choose randomly (e.g. `openssl rand -base64 64`)

**Comment by mhatvan** [comment]
> I think we can stick to yarn, it has great support across the ecosystem and basically already does everything that pnpm does for our use-case. Also could you please stick to the default eslint and prettier configs? They are already decent defaults and at least regarding the formatting part I'd like to keep it that way.

yarn 1 development is frozen, has almost 2k open issues and there were no updates for more than a year. Also you stated that you don't like waiting, pnpm is super fast and efficient. Both /ferdi and/recipes are in the process to switch to pnpm, so this makes it consistent, I don't see any downsides here.

Regarding eslint and prettier configs: all I did was merge latest master into this branch and combine your configs with the ones coming from master. I am not opinionated regarding eslint or formatting style, so feel free to change the configuration however you like and run it over all files.

**Comment by cromefire** [comment]
I removed most non standard stuff from the style configs, some of your changes seemed to have killed some part of the typescript build though.

**Comment by cromefire** [comment]
Should be fixed now for the most part.

**Comment by cromefire** [comment]
Seems like you where faster, thanks
